cps/main.c

```diff
@@ -296,15 +296,57 @@ static void
 process_ingress(struct gatekeeper_if *iface, struct rte_kni *kni,
 	uint16_t rx_queue)
 {
-	struct rte_mbuf *bufs[GATEKEEPER_MAX_PKT_BURST];
-	uint16_t num_rx = rte_eth_rx_burst(iface->id, rx_queue, bufs,
+	struct rte_mbuf *rx_bufs[GATEKEEPER_MAX_PKT_BURST];
+	uint16_t num_rx = rte_eth_rx_burst(iface->id, rx_queue, rx_bufs,
 		GATEKEEPER_MAX_PKT_BURST);
-	unsigned int num_tx = rte_kni_tx_burst(kni, bufs, num_rx);
+	uint16_t num_kni = 0;
+	unsigned int num_tx;
+	uint16_t i;
+
+	if (!iface->vlan_insert) {
+		num_kni = num_rx;
+		goto kni_tx;
+	}
+
+	/* Remove any VLAN headers before passing to the KNI. */
+	for (i = 0; i < num_rx; i++) {
+		struct ether_hdr *eth_hdr =
+			rte_pktmbuf_mtod(rx_bufs[i], struct ether_hdr *);
+		struct vlan_hdr *vlan_hdr;
+
+		RTE_VERIFY(num_kni <= i);
+
+		if (eth_hdr->ether_type != rte_cpu_to_be_16(ETHER_TYPE_VLAN)) {
+			RTE_LOG(WARNING, GATEKEEPER,
+				"cps: %s iface is configured for VLAN but received a non-VLAN packet\n",
+				iface->name);
+			goto to_kni;
+		}
+
+		/* Copy Ethernet header over VLAN header. */
+		vlan_hdr = (struct vlan_hdr *)&eth_hdr[1];
+		eth_hdr->ether_type = vlan_hdr->eth_proto;
+		memmove((uint8_t *)eth_hdr + sizeof(struct vlan_hdr), eth_hdr,
+			sizeof(*eth_hdr));
+
+		/* Remove the unneeded bytes from the front of the buffer. */
+		if (rte_pktmbuf_adj(rx_bufs[i],
+				sizeof(struct vlan_hdr)) == NULL) {
```

Comment:
Given that we properly used `unlikely()` in `if (unlikely(num_kni < i))` below, we should use `unlikely()` here as well since the two tests are related.